pe.c

```diff
@@ -1079,6 +1081,16 @@ handle_image (void *data, unsigned int datasize,
 			return EFI_UNSUPPORTED;
 		}
 
+		efi_status = parse_sbat_var(&sbat_var_entries);
+		/* we get EFI_SUCCESS only if variable is fully parsed,
+		   and EFI_NOT FOUND means do not perform verificaiton
+		   continue with EFI_SUCCESS
+		*/
+		if (efi_status == EFI_SUCCESS)
+			efi_status = verify_sbat(&sbat, &sbat_var_entries);
+		else if (efi_status == EFI_NOT_FOUND)
+			efi_status = EFI_SUCCESS;
```

Comment:
you are missing a else clause here to return errors `if (efi_status != EFI_SUCCESS && efi_status != EFI_NOT_FOUND)`.

Reply:
Actually, probably makes sense to just do `if (EFI_ERROR(efi_status))` here, to catch both `parse_sbat_var()` and `verify_sbat()` returned errors.

Reply:
I am not actually sure i understand you proposal ( sleep is important stuff )
Currently we do verify_sbat only in case parse_sbat_var returned EFI_SUCCESS.

If parse_sbat_var returned anything else, except for EFI_NOT_FOUND and EFI_SUCCESS this whole loop will not be executed, and error will be propagated to a later stages of code.

verify_sbat can also return only:
EFI_INVALID_PARAMETER
EFI_SECURITY_VIOLATION
EFI_SUCCESS

so we will only be getting EFI_NOT_FOUND only in case parse_sbat_var returned it.

The check for (EFI_ERROR(efi_status) should be at the place, where we halt the failed verification.
Currently even after failed sbat we still do verify_buffer, and binary can be validated by it: whenever we want to fail boot immediately after SBAT failure is what i do not know.

Reply:
> f parse_sbat_var returned anything else, except for EFI_NOT_FOUND and EFI_SUCCESS this whole loop will not be executed, and error will be propagated to a later stages of code.

That's not true. If `parse_sbat_var()` returned anything that's not EFI_NOT_FOUND || EFI_SUCCESS, then this error value will not be propagated and `verify_buffer()` wrongly called.

Reply:
My proposal is to do something like the following:

```
               efi_status = parse_sbat_var(&sbat_var_entries);
               /*
                * Until a SBAT variable is installed into the systems, it is expected that
                * attempting to parse the variable will fail with a EFI_NOT_FOUND error.
                *
                * Do not consider that error fatal for now.
                */
               if (efi_status == EFI_SUCCESS)
                       efi_status = verify_sbat(&sbat, &sbat_var_entries);
               else if (efi_status == EFI_NOT_FOUND)
                       efi_status = EFI_SUCCESS;

               if (EFI_ERROR(efi_status)) {
                 // print error log and do any cleanup needed
                 return efi_status;
               }

               efi_status = verify_buffer(data, datasize,
                                           &context, sha256hash, sha1hash);
```

Reply:
> Currently even after failed sbat we still do verify_buffer, and binary can be validated by it: whenever we want to fail boot immediately after SBAT failure is what i do not know.

If we fail to either parse the SBAT variable (i.e: any other error than EFI_NOT_FOUND) or the `.sbat` section, then I think that we can't boot the binary. Otherwise we risk to boot a binary that has been already revoked by SBAT.

So in that case `verify_buffer()` should be cal it and `handle_image()` return earlier IMO.

Reply:
OK, we are talking about same thing. I was thinking about binaries, that might be enrolled to allowlist by hash. We still would fail them. If yes, this pretty much what i was trying to understand.